src/vantage6_strongaya_general/privacy_measures.py

```diff
@@ -0,0 +1,1427 @@
+    different sessions or algorithm runs.
+
+    Args:
+        df_shape: Shape of the input DataFrame (rows, columns)
+        variable: Variable to apply DP to
+        epsilon: Privacy parameter
+        delta: Failure probability parameter
+        distribution: Noise distribution mechanism
+        operation: Statistical operation (for numerical data)
+        **kwargs: Additional parameters affecting the result
+
+    Returns:
+        int: A deterministic seed value for the random number generator
+    """
+    # Create a dictionary of all query parameters
+    query_dict = {
+        "df_shape": df_shape,
+        "variable": variable,
+        "epsilon": epsilon,
+        "delta": delta,
+        "distribution": distribution,
+    }
+
+    # Add operation if provided (for numerical data)
+    if operation:
+        query_dict["operation"] = operation
+
+    # Add additional parameters that affect query results
+    for param in ["top_k", "bin_count"]:
+        if param in kwargs:
+            query_dict[param] = kwargs[param]
+
+    # Convert to JSON string and hash
+    query_json = json.dumps(query_dict, sort_keys=True)
+    hash_value = hashlib.sha256(query_json.encode()).hexdigest()
+
+    # Convert hash to integer for seed (modulo to prevent overflow)
+    return int(hash_value, 16) % (2**32 - 1)
+
+
+def _generate_noise(
+    sensitivity: float,
+    epsilon: float,
+    delta: float = 1e-5,
+    distribution: NoiseDistribution = "laplace",
+    seed: Optional[int] = None,
+) -> float:
+    """
+    Generate noise based on the chosen differential privacy mechanism.
+
+    Using a consistent seed ensures the same query always produces the same noise,
+    preventing multiple query averaging attacks while maintaining privacy guarantees.
+
+    Args:
+        sensitivity (float): The sensitivity of the query.
+        epsilon (float): The privacy parameter (smaller means more privacy).
+        delta (float): The failure probability parameter (used only for Gaussian).
+        distribution (NoiseDistribution): The DP mechanism to use:
+            - 'laplace': Provides pure ε-differential privacy
+            - 'gaussian': Provides (ε,δ)-differential privacy
+        seed (Optional[int]): Seed for the random number generator, used for deterministic noise.
+            If provided, the same seed will always produce the same noise value.
+
+    Returns:
+        float: A noise value from the appropriate distribution.
+    """
+    # Create a deterministic RNG if seed is provided
+    if seed is not None:
+        # Create a separate random number generator that won't affect other code
+        rng = np.random.RandomState(seed)
+    else:
+        # Use the global RNG if no seed (less secure but maintains backwards compatibility)
+        rng = np.random.RandomState()
+
+    if distribution == "laplace":
+        scale = _get_laplace_scale(sensitivity, epsilon)
+        return rng.laplace(0, scale)
+    else:  # gaussian
```

Comment:
consider throwing error if `distribution` is something else but laplace or gaussian